bitacross-worker/bitacross/core/bc-musig2-ceremony/src/lib.rs

```diff
@@ -28,86 +28,72 @@ use std::{format, string::String, sync::Arc};
 
-use k256::SecretKey;
-use musig2::{
-	secp::Point, verify_single, BinaryEncoding, CompactSignature, KeyAggContext, LiftedSignature,
-	SecNonceSpices,
-};
-use std::{vec, vec::Vec};
+#[cfg(feature = "std")]
+use std::sync::RwLock;
+
+#[cfg(feature = "sgx")]
+use std::sync::SgxRwLock as RwLock;
 
-use crate::CeremonyEvent::CeremonyEnded;
 use codec::{Decode, Encode};
 use itp_sgx_crypto::{key_repository::AccessKey, schnorr::Pair as SchnorrPair};
+use k256::SecretKey;
 pub use k256::{elliptic_curve::sec1::FromEncodedPoint, PublicKey};
 use litentry_primitives::RequestAesKey;
-use log::{debug, error, info, trace};
-use musig2::secp::Scalar;
+use log::*;
+use musig2::{
+	secp::{Point, Scalar},
+	verify_single, BinaryEncoding, CompactSignature, KeyAggContext, LiftedSignature,
+	SecNonceSpices,
+};
 pub use musig2::{PartialSignature, PubNonce};
-use std::collections::HashMap;
+use std::{
+	collections::HashMap,
+	time::{SystemTime, UNIX_EPOCH},
+	vec,
+	vec::Vec,
+};
 
 pub type CeremonyId = SignBitcoinPayload;
 pub type SignaturePayload = Vec<u8>;
 pub type Signers = Vec<SignerId>;
-pub type CeremonyRegistry<AK> = HashMap<CeremonyId, MuSig2Ceremony<AK>>;
-pub type CeremonyCommandsRegistry = HashMap<CeremonyId, Vec<PendingCeremonyCommand>>;
+pub type CeremonyRegistry<AK> = HashMap<CeremonyId, (Arc<RwLock<MuSig2Ceremony<AK>>>, u64)>;
+pub type CeremonyCommandTmp = HashMap<CeremonyId, (Arc<RwLock<Vec<CeremonyCommand>>>, u64)>;
 // enclave public key is used as signer identifier
 pub type SignerId = [u8; 32];
 pub type SignersWithKeys = Vec<(SignerId, PublicKey)>;
 
-pub struct PendingCeremonyCommand {
-	pub ticks_left: u32,
-	pub command: CeremonyCommand,
-}
-
-impl PendingCeremonyCommand {
-	pub fn tick(&mut self) {
-		self.ticks_left -= 1;
-	}
-}
-
-#[derive(Encode, Debug)]
-pub enum SignBitcoinError {
-	InvalidSigner,
-	CeremonyError,
-}
-
-#[derive(Debug, Eq, PartialEq)]
+#[derive(Debug, Eq, PartialEq, Encode)]
 pub enum CeremonyError {
-	NonceReceivingError(NonceReceivingErrorReason),
-	PartialSignatureReceivingError(PartialSignatureReceivingErrorReason),
+	CeremonyInitError(CeremonyErrorReason),
+	NonceReceivingError(CeremonyErrorReason),
+	PartialSignatureReceivingError(CeremonyErrorReason),
 }
 
-#[derive(Debug, Eq, PartialEq)]
-pub enum NonceReceivingErrorReason {
+#[derive(Debug, Eq, PartialEq, Encode)]
+pub enum CeremonyErrorReason {
+	AlreadyExist,
+	CreateCeremonyError,
 	SignerNotFound,
 	ContributionError,
 	IncorrectRound,
-	FirstRoundFinalizationError,
+	RoundFinalizationError,
 }
 
-#[derive(Debug, Eq, PartialEq)]
-pub enum PartialSignatureReceivingErrorReason {
-	SignerNotFound,
-	ContributionError,
-	IncorrectRound,
-	SecondRoundFinalizationError,
-}
-
-// events come from outside and are consumed by ceremony in tick fn
-#[derive(Debug)]
+#[derive(Debug, PartialEq, Eq, Clone)]
 pub enum CeremonyCommand {
+	InitCeremony([u8; 32], SignersWithKeys, bool),
```

Comment:
It looks like this command is incomplete - it lacks of `SignBitcoinPayload`. 
Could you add it for the sake of simplicity?

Reply:
`SignBitcoinPayload` is using by all kinds of commands, so I put it together with command in a tuple `Option<(CeremonyId, CeremonyCommand)>`.
If `InitCeremony` add `SignBitcoinPayload` then all variants should do so, and we need another function for taking out `SignBitcoinPayload` from variants.

Reply:
Updated